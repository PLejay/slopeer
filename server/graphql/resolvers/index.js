const query = require('./query.resolver');
const mutation = require('./mutation.resolver');
const types = require('./types.resolver');

module.exports = {
  Query: {
    routes: query.routes,
    route: query.route,
    user: query.user
  },
  Mutation: {
    createRoute: mutation.createRoute,
    updateRoute: mutation.updateRoute,
    removeRoute: mutation.removeRoute,
    createUser: mutation.createUser,
    updateUser: mutation.updateUser,
    saveRoute: mutation.saveRoute,
    unsaveRoute: mutation.unsaveRoute,
    login: mutation.login
  },
  User: {
    owned_routes: types.get_owned_routes,
    saved_routes: types.get_saved_routes
  },
  Route: {
    author: types.get_author
  }
};
