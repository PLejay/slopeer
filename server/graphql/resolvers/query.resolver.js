const User = require('../../models/user.model');
const Route = require('../../models/route.model');

exports.routes = async (_, args) =>
  await Route.find({ ...args })
    .populate('author');

exports.route = () => { };

exports.user = async (_, { _id }) =>
  await User.findById(_id)
    .populate('saved_routes')
    .populate('owned_routes');
