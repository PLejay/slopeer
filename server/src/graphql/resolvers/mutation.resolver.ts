import bcrypt from 'bcrypt';
import { Response } from 'express';
import User, { Iuser } from '../../models/user.model';
import Route, { Iroute } from '../../models/route.model';
const { uploadProfilePicture, uploadRoutePicture } = require('../../utils/uploads');
import { ID } from '../../types/types';

type Input = {
  _id?: ID;
  input: Iroute;
  routeId?: ID;
  userId?: ID;
}

type Res = {
  res: Response;
}

export const createRoute = async (_: any, { input }: Input) => {
  const route = new Route({ ...input, picture: null });

  if (input.picture) {
    const picturePath = await uploadRoutePicture(input.picture, route._id);
    route.picture = picturePath;

  }

  await route.save();
  await User.findByIdAndUpdate(input.author, { $push: { 'owned_routes': String(route._id) } },
    { useFindAndModify: false })
  return route;
};

export const updateRoute = async (_: any, { _id, input }: Input) => {
  if (input.picture) {
    const picturePath = await uploadRoutePicture(input.picture, _id);
    input.picture = picturePath;
  }
  return await Route.findByIdAndUpdate(_id, input, { new: true, useFindAndModify: false });
};

exports.removeRoute = async (_: any, { _id }: Input) => {
  const route = await Route.findByIdAndDelete(_id);
  if (route) await User.findByIdAndUpdate(route.author, { $pull: { 'owned_routes': _id } }, { useFindAndModify: false });
  await User.updateMany({}, { $pull: { 'saved_routes': _id } });
  return route;
};
type createUserQuery = {
  input: { email: string; username: string; password: string }
}
export const createUser = async (_: any, { input: { email, username, password } }: createUserQuery, { res }: Res) => {
  let user = await User.findOne({ email });
  if (user) {
    res.status(409);
    return;
  }

  user = new User({ email, username, password, owned_routes: [], saved_routes: [] });
  const salt = await bcrypt.genSalt(10);
  user.password = await bcrypt.hash(user.password, salt);
  await user.save();

  return user.generateAuthToken();
};

exports.updateUser = async (_: any, { _id, input }: any) => {
  if (input.profile_picture) {
    const picturePath = await uploadProfilePicture(input.profile_picture, _id);
    input.profile_picture = picturePath;
  }
  return await User.findByIdAndUpdate(_id, input, { new: true, useFindAndModify: false });
};

exports.saveRoute = async (_: any, { userId, routeId }: Input) =>
  await User.findByIdAndUpdate(userId, { $push: { 'saved_routes': routeId } }, { new: true, useFindAndModify: false });


exports.unsaveRoute = async (_: any, { userId, routeId }: Input) =>
  await User.findByIdAndUpdate(userId, { $pull: { 'saved_routes': routeId } }, { new: true, useFindAndModify: false });

exports.login = async (_: any, { email, password }: Iuser, { res }: Res) => {
  const user = await User.findOne({ email });
  if (!user) {
    res.status(400);
    return;
  }
  const validPassword = await bcrypt.compare(password, user.password);
  if (!validPassword) return res.status(400);

  return user.generateAuthToken();
};
