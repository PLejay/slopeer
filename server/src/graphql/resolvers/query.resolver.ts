import  User  from '../../models/user.model';
import Route  from '../../models/route.model';



export const routes = async (_:any, args:[string]) =>
  await Route.find({ ...args })
    .populate('author');

export const route = async (_:any, { _id }) => await Route.findById(_id).populate('author');

export const user = async (_:any, { _id }) =>
  await User.findById(_id)
    .populate('saved_routes')
    .populate('owned_routes');
