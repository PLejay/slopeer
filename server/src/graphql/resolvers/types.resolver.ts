import { Iroute } from "../../models/route.model";

const { User } = require('../../models');

const getRoutes = async (_id, routes: string) => {
  const user = await User.findById(_id).populate(routes);
  return user:Iuser[routes];
};

exports.get_owned_routes = async ({ _id }) => await getRoutes(_id, 'owned_routes');

exports.get_saved_routes = async ({ _id }) => await getRoutes(_id, 'saved_routes');

exports.get_author = async ({ author: id }) => await User.findById(id);
