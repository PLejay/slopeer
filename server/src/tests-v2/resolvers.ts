const users = [...Array(5).keys()].map(key => ({
  _id: key + '',
  username: `Name${key}`,
  email: `Name${key}@mail.com`,
  password: '123'
}));

type Input = {
  _id: string;
}

export default {
  Query: {
    findUser: (_: any, { _id }: Input) => {
      const user = users.find(user => user._id === _id);
      if (user) {
        return user;
      } else {
        throw new Error('User not found!')
      }
    }
  },

  Mutation: {
    deleteUser: (_: any, { _id }: Input) => {
      const index = users.findIndex(user => user._id === _id);
      if (index < 0) return false;
      users.splice(index, 1);
      return true;
    }
  }
}