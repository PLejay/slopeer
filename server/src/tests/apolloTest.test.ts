import {ApolloServer, gql } from  'apollo-server-express';
import {Request, Response} from 'express';
import resolvers from '../graphql/resolvers'
import typeDefs from '../graphql/schemas'
import {createTestClient} from 'apollo-server-testing'
import Route from '../models/route.model';

// route with just required elements
const mockRoute = {
_id: '1',
name: 'a route',
grade: 'B1',
author: '001',
lat: '0023.023.23',
lng: '2100.103.90'
}

//mock db content 
jest.spyOn(Route, 'find')
      .mockImplementationOnce(() => {
        return {populate: ()=> [mockRoute] }
      }) //mock find method

type express= {res:Response, req: Request}  

export const server = new ApolloServer({
  typeDefs,
  resolvers,
  context: ({ req, res }:express) => ({ req, res }),
  uploads: false
  
  // mocks:true,
  // mockEntireSchema: false,
});

const GET_ROUTES = gql`
query{
  routes(public: true) {
    _id
    grade
    lat
    lng
  }
}`

const {query} = createTestClient(server as any);


describe('route query', ()=>{

  it('is able to get routes', async ()=> {
    const res = await query({query:GET_ROUTES})
    console.log(res.data.routes, 'res')
   
    //snapshot shows that the data is there, but cant probe anything else
    expect(res).toMatchSnapshot();
     //as the query is just getting 4 of the elements that are part of route it does not pass a test toEqual
     expect(res.data.routes).toEqual([mockRoute])
  })
  //can not make two at the same time, error
  // it('should get routes', async ()=> {
  //   const res = await query({query:GET_ROUTES})
  //   console.log(res, 'res')
  //   
  // })


})