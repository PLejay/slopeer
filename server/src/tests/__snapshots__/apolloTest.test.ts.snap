// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`route query is able to get routes 1`] = `
Object {
  "data": Object {
    "routes": Array [
      Object {
        "_id": "1",
        "grade": "B1",
        "lat": "0023.023.23",
        "lng": "2100.103.90",
      },
    ],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
